### PR TITLE
HOT-2482 - Tighten CodeClimate duplication detection

### PR DESCRIPTION
**Pull request by wmitchell-cambria**
### Jira Story

- [Tighten JavaScript code duplication threshold to 120 HOT-2482](https://osi-cwds.atlassian.net/browse/HOT-2482)

## Description

This reduces our CodeClimate JavaScript duplication threshold from 170 to 120. (The default is 40, if we can ever get there.) It contains only refactoring, no new features or functionality. (The biggest/only visual change is capitalizing a letter on an error page.)

I recommend reviewing this on a commit-by-commit basis, since these affect a wide variety of files within the codebase.

## Tests
- [x] I have included unit tests 
- [ ] I have included feature tests 
- [ ] I have included other tests 
- [ ] I have NOT included tests 

This should pass all existing feature tests.

## Types of changes

- [ ] New feature (non-breaking change which adds functionality)
- [x] Refactoring (No behavioral changes)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist:

- [x] My code follows the code style of this project.
- [x] I have ran all tests for the project.
- [x] I have ran lint/rubocop check for the changed/new files.
- [x] My code is in a stable state ready to be deployable, but not necessarily complete.
- [x] I promise on my honor as a CWDS developer to ensure I don't break the build, to check Lint/Rubocop, use best practices, to leave the code in better shape than I found it, and to have a good day.

### CODE REVIEW

**Comment by wmitchell-cambria** [comment]
Jenkins, retest this please. I know I've given you a lot of builds today, but just this one more.

File: app/javascript/common/relationship/CreateRelationshipForm.jsx [hunk 1]
```diff
@@ -74,26 +75,8 @@ PersonInfoList.propTypes = {
...
-    person: PropTypes.shape({
-      age: PropTypes.string,
-      dateOfBirth: PropTypes.string,
-      gender: PropTypes.string,
-      id: PropTypes.string,
-      legacy_id: PropTypes.string,
-      name: PropTypes.string,
-    }),
-  })),
+  candidates: CandidatesPropType,
```
Comment (mlipardo) [inline app/javascript/common/relationship/CreateRelationshipForm.jsx]: 💯

File: app/javascript/data/candidates.js [hunk 2]
```diff
@@ -0,0 +1,20 @@
+import PropTypes from 'prop-types'
```
Comment (mlipardo) [inline app/javascript/data/candidates.js]: 👍

File: app/javascript/reducers/screeningDecisionFormReducer.js [hunk 3]
```diff
@@ -6,44 +6,27 @@ import {
...
-    } = screening
-    return fromJS({
-      screening_decision: untouched(screening_decision),
-      screening_decision_detail: untouched(screening_decision_detail),
-      screening_contact_reference: untouched(screening_contact_reference),
-      access_restrictions: untouched(access_restrictions),
-      restrictions_rationale: untouched(restrictions_rationale),
-      additional_information: untouched(additional_information),
-    })
+    return toUntouchedObject(FIELDS, screening)
```
Comment (mlipardo) [inline app/javascript/reducers/screeningDecisionFormReducer.js]: +1

File: app/javascript/reducers/screeningDecisionFormReducer.js [hunk 4]
```diff
@@ -6,44 +6,27 @@ import {
...
-  }}) {
-    return state.setIn(['screening_decision', 'value'], screening_decision)
-      .setIn(['screening_decision_detail', 'value'], screening_decision_detail)
-      .setIn(['screening_contact_reference', 'value'], screening_contact_reference)
-      .setIn(['access_restrictions', 'value'], access_restrictions)
-      .setIn(['restrictions_rationale', 'value'], restrictions_rationale)
-      .setIn(['additional_information', 'value'], additional_information)
+  [RESET_SCREENING_DECISION_FIELD_VALUES](state, {payload}) {
+    return FIELDS.reduce(
+      (newState, field) => newState.setIn([field, 'value'], payload[field]),
```
Comment (mlipardo) [inline app/javascript/reducers/screeningDecisionFormReducer.js]: and this one too 👍

File: app/javascript/errors/ErrorPage.jsx [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+    </div>
+  </div>
+)
+
+ErrorPage.propTypes = {
+  details: PropTypes.string.isRequired,
+  message: PropTypes.string.isRequired,
+}
+
+export default ErrorPage
```
Comment (NagaPradeepGajula-osi) [inline app/javascript/errors/ErrorPage.jsx]: 💯

File: spec/javascripts/components/errors/NotFoundPageSpec.jsx [hunk 6]
```diff
@@ -1,12 +1,12 @@
...
 import NotFoundPage from 'errors/NotFoundPage'
-import {shallow} from 'enzyme'
+import {mount} from 'enzyme'
 
 describe('NotFoundPage', () => {
   let component
 
   beforeEach(() => {
-    component = shallow(<NotFoundPage/>, {disableLifecycleMethods: true})
+    component = mount(<NotFoundPage/>, {disableLifecycleMethods: true})
```
Comment (NagaPradeepGajula-osi) [inline spec/javascripts/components/errors/NotFoundPageSpec.jsx]: why we are changing it to mount here? any specific reason

**Comment by NagaPradeepGajula-osi** [comment]
Nit: may be you could have squashed the last three commits into one.

File: spec/javascripts/components/errors/NotFoundPageSpec.jsx (hunk 6 above)
Comment (wmitchell-cambria) [reply spec/javascripts/components/errors/NotFoundPageSpec.jsx]: Depending on how you look at it either:
* Changing to `mount` allowed me to preserve more of the existing unit tests, minimizing the risk of breaking something.
* I was lazy and didn't want to change all of the tests.

Now that I have actually looked at the error pages manually, I can circle back, update the tests, and change it back to shallow. I don't think the performance cost of this is too bad, though, since they are still relatively small components overall.

**Comment by wmitchell-cambria** [comment]
kk, I was tired of Pradeep calling me out on my bad git history, so I fixed the ErrorPage tests, squashed related commits, and rebased on master. Apologies if that messed up any current reviewers... :-(

**Comment by wmitchell-cambria** [comment]
I know, but it was bugging me, too. :-D

**Comment by NagaPradeepGajula-osi** [comment]
Just kidding. But thanks for rebasing. 👍

**Comment by wmitchell-cambria** [comment]
Since it came up, and I want a place to point to if it happens to come up again, these types of refactors create two interesting dilemmas with unit tests.

First, if you refactor a function from being used internally in two modules to being exported by a shared module, our typical practice is that the newly exported method should be unit tested. This creates extra work for the person doing the refactoring, which means fewer people will do these refactors day to day. One argument would be that the newly exported function is already tested by the existing unit tests for the modules where it is used, and adding more unit tests probably just duplicates existing ones. However, since it is exported, there is nothing stopping a third module from using it, and it should have unit tests to guarantee its behavior in that situation. Since unit tests run with pretty low overhead, having the tests in the original modules stick around as 'extra' seems low cost as well. Verdict: Write unit tests for newly exported functions.

A second dilemma is more interesting... Factoring out React components in the same way. If you factor JSX that is used internally into a separate React component, and the original component was shallow rendered, you have a choice. Fully mount the original component so that it has access to the full render tree as it did before, or keep it as a shallow render and potentially drastically change the tests for that component. The latter is better for performance but has a higher chance of introducing a behavioral change (bug) in the refactoring process. I do not have a great answer for this, except maybe to test that it passes the original tests first, and *only then* change the tests to use shallow rendering in a second step. It is potentially worth noting that this problem is lessened if our components are small and composable to begin with.

Anyways, I did a mix of both in this PR, and so I wanted to jot down my thoughts for later.
